# Exploration_2

* 아래에서 다룰 3가지 데이터셋 문제는 모두 scikit-learn에 내장된 예제 데이터셋으로 정답을 맞히기 위해 학습하는 지도학습(Supervised Learning)이다.
* 지도학습은 다시 분류(classification)과 회귀(regression)으로 나눌 수 있다.
* 분류는 입력받은 데이터를 특정 카테고리 중 하나로 분류해내는 문제를, 회귀는 입력받은 데이터에 따라 필드의 수치를 맞히는 문제를 말한다.
* 아래의 데이터셋은 모두 분류 문제로 5개의 분류 모델을 학습시키고 예측해보자.

## 1. load_digits()

### (1) 필요한 모듈 import하기

In [1]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
import pandas as pd

* 사이킷런(sicit-learn)은 파이썬을 기반으로 한 머신러닝 라이브러리
* TensorFlow, PyTorch등 딥러닝에 특화된 라이브러리들이 강세롤 보이지만, 머신러닝의 다양한 알고리즘과 편리한 프레임워크를 제공한다는 점으로 인해 여전히 많은 데이터 분석가들이 사용.
---
#### 아래에서 사용할 분류 모델들
1. Decision Tree : 의사 결정을 할, 즉 데이터를 분리할 어떤 경계를 찾아내어 데이터를 체에 거르듯 한 단계씩 분류해나가는 모델이다. 직관적이면서도 간단하게 사용할 수 있어 분류 문제를 풀 때 가장 기본적으로 쓰이는 모델 중 하나이다. 의사결정나무는 결정경계(decision boundary)가 데이터 축에 수직이어서 특정 데이터에만 잘 작동할 가능성이 높다. 이를 극복하기 위해 제안된 모델이 Random Forest이며, 여러 개의 Decision Tree를 합쳐서 만들어놓은 개념이다.
2. Random Forest : 여러 개의 의사 결정 트리를 모아 놓은 것으로, 각각의 의사 결정 트리를 만들기 위해 쓰이는 특성들을 랜덤으로 선택한다. 상위 모델들이 예측하는 편향된 결과보다, 다양한 모델들의 결과를 반영함으로써 더 다양한 데이터에 대한 의사결정을 내릴 수 있게 한다. 
3. SVM(Support Vector Machine) : 결정경계(Decision Boundary), 즉 분류를 위한 기준 선을 정의하는 모델이다. 그래서 분류되지 않은 새로운 점이 나타나면 경계의 어느 쪽에 속하는지 확인하여 분류 과제를 수행한다. 
4. SGD(Stochastic Gradient Descent) : 결과값과 실제값 사이의 차이를 정의하는 loss function(손실함수)의 값을 최소화하기 위해 기울기를 이용하는 모델이다.
5. Logistic Regression : 독립변수의 선형 결합을 이용하여 사건의 발생 가능성을 예측하는데 사용되는 통계 기법 모델이다. 선형회귀 분석과는 다르게 종속 변수가 범주형 데이터를 대항으로 하며 입력 데이터가 주어졌을 때 해당 데이터의 결과가 특정 분류로 나뉘기 때문에 일종의 분류 기법으로 볼 수 있다.

### (2) 데이터 준비

In [2]:
# digits이라는 변수에 데이터를 담아준다.
digits = load_digits()

### (3) 데이터 이해하기

In [3]:
# 데이터에 어떤 정보들이 담겨있는지 훓어보자!
digits

{'data': array([[ 0.,  0.,  5., ...,  0.,  0.,  0.],
        [ 0.,  0.,  0., ..., 10.,  0.,  0.],
        [ 0.,  0.,  0., ..., 16.,  9.,  0.],
        ...,
        [ 0.,  0.,  1., ...,  6.,  0.,  0.],
        [ 0.,  0.,  2., ..., 12.,  0.,  0.],
        [ 0.,  0., 10., ..., 12.,  1.,  0.]]),
 'target': array([0, 1, 2, ..., 8, 9, 8]),
 'frame': None,
 'feature_names': ['pixel_0_0',
  'pixel_0_1',
  'pixel_0_2',
  'pixel_0_3',
  'pixel_0_4',
  'pixel_0_5',
  'pixel_0_6',
  'pixel_0_7',
  'pixel_1_0',
  'pixel_1_1',
  'pixel_1_2',
  'pixel_1_3',
  'pixel_1_4',
  'pixel_1_5',
  'pixel_1_6',
  'pixel_1_7',
  'pixel_2_0',
  'pixel_2_1',
  'pixel_2_2',
  'pixel_2_3',
  'pixel_2_4',
  'pixel_2_5',
  'pixel_2_6',
  'pixel_2_7',
  'pixel_3_0',
  'pixel_3_1',
  'pixel_3_2',
  'pixel_3_3',
  'pixel_3_4',
  'pixel_3_5',
  'pixel_3_6',
  'pixel_3_7',
  'pixel_4_0',
  'pixel_4_1',
  'pixel_4_2',
  'pixel_4_3',
  'pixel_4_4',
  'pixel_4_5',
  'pixel_4_6',
  'pixel_4_7',
  'pixel_5_0',
  'pixel_5_1',
 

In [4]:
type(digits) # Bunch는 파이썬의 딕셔너리와 유사한 형테의 데이터 타입

sklearn.utils.Bunch

In [5]:
# digit data의 key값들을 불러와서 살펴보기
digits.keys() 

dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])

In [6]:
digit_data = digits.data
digit_data

array([[ 0.,  0.,  5., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ..., 10.,  0.,  0.],
       [ 0.,  0.,  0., ..., 16.,  9.,  0.],
       ...,
       [ 0.,  0.,  1., ...,  6.,  0.,  0.],
       [ 0.,  0.,  2., ..., 12.,  0.,  0.],
       [ 0.,  0., 10., ..., 12.,  1.,  0.]])

In [7]:
digit_data.shape 
# shape은 배열의 형상정보를 출력
# 총 1797개의 데이터를 가지고 있고 각각 64개의 정보를 담고있다.

(1797, 64)

In [8]:
# feature data 지정하기
digit_feature = digits.feature_names
digit_feature
# 64개의 정보는 'pixel_0_0', 'pixel_0_1'...와 같은 카테코리를 의미한다.

['pixel_0_0',
 'pixel_0_1',
 'pixel_0_2',
 'pixel_0_3',
 'pixel_0_4',
 'pixel_0_5',
 'pixel_0_6',
 'pixel_0_7',
 'pixel_1_0',
 'pixel_1_1',
 'pixel_1_2',
 'pixel_1_3',
 'pixel_1_4',
 'pixel_1_5',
 'pixel_1_6',
 'pixel_1_7',
 'pixel_2_0',
 'pixel_2_1',
 'pixel_2_2',
 'pixel_2_3',
 'pixel_2_4',
 'pixel_2_5',
 'pixel_2_6',
 'pixel_2_7',
 'pixel_3_0',
 'pixel_3_1',
 'pixel_3_2',
 'pixel_3_3',
 'pixel_3_4',
 'pixel_3_5',
 'pixel_3_6',
 'pixel_3_7',
 'pixel_4_0',
 'pixel_4_1',
 'pixel_4_2',
 'pixel_4_3',
 'pixel_4_4',
 'pixel_4_5',
 'pixel_4_6',
 'pixel_4_7',
 'pixel_5_0',
 'pixel_5_1',
 'pixel_5_2',
 'pixel_5_3',
 'pixel_5_4',
 'pixel_5_5',
 'pixel_5_6',
 'pixel_5_7',
 'pixel_6_0',
 'pixel_6_1',
 'pixel_6_2',
 'pixel_6_3',
 'pixel_6_4',
 'pixel_6_5',
 'pixel_6_6',
 'pixel_6_7',
 'pixel_7_0',
 'pixel_7_1',
 'pixel_7_2',
 'pixel_7_3',
 'pixel_7_4',
 'pixel_7_5',
 'pixel_7_6',
 'pixel_7_7']

In [9]:
# label data 지정하기
# 머신러닝 모델에게 특성(features)을 입력했을때, 0~9 숫자 중 하나를 출력하도록 학습시켜야 한다.
# 이렇게 머신러닝 모델이 출력해야 하는 정답을 라벨(label) 또는 타겟(target)이라고 한다.
digit_label = digits.target
print(digit_label.shape)
digit_label

(1797,)


array([0, 1, 2, ..., 8, 9, 8])

In [10]:
# target names 출력해보기
digits.target_names # 0~9까지 숫자

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [11]:
# 데이터 describe해보기
# DESCR : 데이터셋의 설명
print(digits.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

In [12]:
# pandas 라이브러리는 파이썬에서 표 형태로 이루어진 2차원 배열 데이터를 다루는데 가장 많이 쓰이는 도구
# 표 데이터를 활용해서 데이터 분석을 하기도 하고, 대형 데이터의 여러 통계량을 다루는 것에도 최적화
# digits 데이터 또한 행과 열이 있는 2차원 데이터이므로 pandas를 활용해서 다뤄보았다.
# 숫자 데이터셋을 pandas가 제공하는 DataFrame 이라는 자료형으로 변환해보았다.
# 정답 데이터도 함께 있다면 데이터를 다루기 더 편리
# ['label']이라는 코드로 라벨 컬럼 추가
digit_df = pd.DataFrame(data = digit_data, columns = digits.feature_names)
digit_df["label"] = digits.target
digit_df
# 64가지 feature데이터들은 바로 머신러닝 모델이 풀어야 하는 문제지
# 0,1,2, ..., 9와 같이 표현된 label데이터는 머신러닝 모델에게 정답지
# 문제지 : 머신러닝 모델에게 입력되는 데이터. feature라고 부르기도 함. 변수 이름으로는 X를 많이 사용.
# 정답지 : 머신러닝 모델이 맞혀야 하는 데이터. label 또는 target이라고 함. 변수 이름으로는 y를 많이 사용.

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0           0.0        0.0        5.0       13.0        9.0        1.0   
1           0.0        0.0        0.0       12.0       13.0        5.0   
2           0.0        0.0        0.0        4.0       15.0       12.0   
3           0.0        0.0        7.0       15.0       13.0        1.0   
4           0.0        0.0        0.0        1.0       11.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        4.0       10.0       13.0        6.0   
1793        0.0        0.0        6.0       16.0       13.0       11.0   
1794        0.0        0.0        1.0       11.0       15.0        1.0   
1795        0.0        0.0        2.0       10.0        7.0        0.0   
1796        0.0        0.0       10.0       14.0        8.0        1.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_7  pixel_7_0  \
0           0.0        0.0        0.0        0.0  ...        0.0        0.0   
1           0.0        0.0        0.0        0.0  ...        0.0        0.0   
2           0.0        0.0        0.0        0.0  ...        0.0        0.0   
3           0.0        0.0        0.0        8.0  ...        0.0        0.0   
4           0.0        0.0        0.0        0.0  ...        0.0        0.0   
...         ...        ...        ...        ...  ...        ...        ...   
1792        0.0        0.0        0.0        1.0  ...        0.0        0.0   
1793        1.0        0.0        0.0        0.0  ...        0.0        0.0   
1794        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1795        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1796        0.0        0.0        0.0        2.0  ...        0.0        0.0   

      pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  pixel_7_6  \
0           0.0        6.0       13.0       10.0        0.0        0.0   
1           0.0        0.0       11.0       16.0       10.0        0.0   
2           0.0        0.0        3.0       11.0       16.0        9.0   
3           0.0        7.0       13.0       13.0        9.0        0.0   
4           0.0        0.0        2.0       16.0        4.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        2.0       14.0       15.0        9.0        0.0   
1793        0.0        6.0       16.0       14.0        6.0        0.0   
1794        0.0        2.0        9.0       13.0        6.0        0.0   
1795        0.0        5.0       12.0       16.0       12.0        0.0   
1796        1.0        8.0       12.0       14.0       12.0        1.0   

      pixel_7_7  label  
0           0.0      0  
1           0.0      1  
2           0.0      2  
3           0.0      3  
4           0.0      4  
...         ...    ...  
1792        0.0      9  
1793        0.0      0  
1794        0.0      8  
1795        0.0      9  
1796        0.0      8  

[1797 rows x 65 columns]

### (4) train, test 데이터 분리

In [13]:
# 학습에 사용하는 training dataset과 
# 모델의 성능을 평가하는 데 사용하는 test dataset으로 데이터셋을 나누는 작업이 필요.
# 데이터셋 분리는 scikt-learn이 제공하는 train_test_split이라는 함수로
# sklearn.model_selection패키지의 train_test_split을 활용하여,
# training dataset과 test dataset을 간단히 분리해보았다.
X_train, X_test, y_train, y_test = train_test_split(digit_data,
                                                   digit_label,
                                                   test_size=0.2,
                                                   random_state=7)
# digit_data는 문제지, 즉 feature. 모델이 숫자를 맞추기 위해 입력받는 특징 데이터. 64가지 특징정보
# digit_label은 모델이 맞추어야 하는 정답값, 즉 label. 0~9까지 숫자
# test_size로는 test dataset의 크기를 조절. 0.2는 전체의 20%를 데스트 데이터로 사용. 일반적으로도 8:2비율을 선호.
# random_state는 train데이터와 test데이터를 분리하는데 적용되는 랜덤성을 결정
print('X_train 개수: ',len(X_train),', X_test 개수: ', len(X_test))

X_train 개수:  1437 , X_test 개수:  360


### (5) 다양한 모델로 학습시켜보기

### 1. Decision Tree

In [14]:
# 첫 번째로 Decision Tree모델을 사용해보자
# Decision tree는 sklearn.tree패키지 안에 DecisionTreeClassifier라는 이름으로 내장되어있음.
# 모델을 import해서 가져오고, decision_tree라는 변수에 모델을 저장해보자.
decision_tree = DecisionTreeClassifier(random_state=30)

# 모델학습
decision_tree.fit(X_train, y_train)
# fit메서드 : training dataset으로 모델을 학습시킨다는 것, training dataset에 맞게 모델을 fitting하는 것
# training dataset에 있는 데이터들을 통해 어떠한 패턴을 파악하고, 그 패턴에 맞게 예측을 할 수 있도록 학습되기 때문
# 모델은 training dataset에 존재하지 않는 데이터에 대해서는 정확한 정답 카테고리가 무엇인지 알지 못한다.
# 다만 training dataset을 통해 학습한 패턴으로 새로운 데이터가 어떤 카테고리에 속할지 예측할 뿐
# 그렇기 때문에 새로운 데이터에 대해서도 잘 맞출 수 있기 위해서는 training dataset이 어떻게 구성되어 있는지가 매우 중요.
# 더 다양한, 더 일반화 된 데이터로 학습이 될수록 새로운 데이터에 대해서도 잘 맞출 수 있는 것

# 모델로 학습이 완료되었으니 test 데이터로 예측해보자
y_pred = decision_tree.predict(X_test)
# x_test데이터에는 정답인 label이 없고 feature데이터만 존재
# 따라서 학습이 완료된 decision_tree모델에 x_test데이터로 
# predict를 실행하면 모델이 예측한 y_pred을 얻게 된다.

print(classification_report(y_test, y_pred))
# 예측한 결과에 대한 수치를 조금 더 편리하게 확인할 수 있는 방법
# scikit-learn에서 성능 평가에 대한 함수들이 모여있는 sklearn.metrics패키지 이용
# 성능을 평가하는 다양한 척도를 classification_report로 간단히 확인해 보자

              precision    recall  f1-score   support

           0       1.00      0.95      0.98        43
           1       0.80      0.79      0.80        42
           2       0.80      0.80      0.80        40
           3       0.89      0.91      0.90        34
           4       0.74      0.95      0.83        37
           5       0.87      0.96      0.92        28
           6       0.87      0.93      0.90        28
           7       0.93      0.79      0.85        33
           8       0.84      0.63      0.72        43
           9       0.71      0.78      0.75        32

    accuracy                           0.84       360
   macro avg       0.85      0.85      0.84       360
weighted avg       0.85      0.84      0.84       360



### 2. Random Forest

In [15]:
# Random Forest : Decision Tree모델을 여러 개 합쳐놓음으로써 그 단점을 극복한 모델.
# 이러한 기법을 앙상블(Ensemble)기법이라고 한다.(데이터 사이언스에서는 의견을 통합하거나 여러가지 결과를 합치는 방식)
# 단일 모델을 여러 개 사용하는 방법을 취함으로써 모델 한 개만 사용할 때의 단점을 집단지성으로 극복하는 개념
# Random Forest는 각각의 의사결정트리를 만드는데 있어 쓰이는 요소들을 무작위적으로 선정
# 총 feature 중 랜덤으로 n개의 feature만 선택해서 하나의 결정트리 생성. 이런식으로 여러 개의 결정트리 생성.
# 하나의 거대한 결정 트리를 만드는 것이 아니라 여러 개의 작은 결정 트리를 만드는 것
# 분류:여러 개의 작은 결정 트리가 예측한 값들 중 가장 많은 값
# 회귀:평균
# 상위 모델들이 예측하는 편향된 결과보다, 다양한 모델들의 결과를 반영함으로써 더 다양한 데이터에 대한 의사결정을 내릴 수 있다.

random_forest = RandomForestClassifier(random_state=20)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.93      1.00      0.97        42
           2       1.00      1.00      1.00        40
           3       0.97      1.00      0.99        34
           4       0.92      0.97      0.95        37
           5       0.90      1.00      0.95        28
           6       1.00      0.96      0.98        28
           7       0.94      0.97      0.96        33
           8       0.97      0.81      0.89        43
           9       0.97      0.94      0.95        32

    accuracy                           0.96       360
   macro avg       0.96      0.96      0.96       360
weighted avg       0.96      0.96      0.96       360



### 3. SVM

In [16]:
# Support Vector Machine(SVM)
# 분류 과제에 사용할 수 있는 강력한 머신러닝 지도학습 모델
# 분류를 위한 기준 선(결정 경계)을 정의하는 모델
# 속성이 3개로 늘어난다면 결정 경계는 '선'이 아닌 '평면'이 됨
# 차원, 즉 속성의 개수가 늘어날수록 결정 경계도 단순한 평면이 아닌 고차원이 되는데 이를 '초평면'이라고 함
# 분류되지 않은 새로운 점이 나타나면 경계의 어느 쪽에 속하는지 확인해서 분류 과제를 수행
# support vector와 hyperplane(초평면)을 이용해서 분류를 수행하게되는 대표적인 선형 분류 알고리즘
# 2차원 공간에서, 즉 데이터에 2개의 클래스만 존재할 때,
# decision boundary(결정 경계):두 개의 클래스를 구분해 주는 선
# 결정 경계는 데이터 군으로부터 최대한 멀리 떨어지는게 좋다.
# support vector:decision boundary에 가까이 있는 데이터
# n개의 속성을 가진 데이터에는 최소 n+1개의 서포트벡터가 존재
# margin:decision boundary와 support vector 사이의 거리
# margin이 넓을수록 새로운 데이터를 잘 구분할 수 있다.(margin최대화->robustness최대화, 최적의 결정 경계)
# SVM 장점 : 대부분의 머신러닝 지도학습 알고리즘은 학습 데이터 모두를 사용하여 모델을 학습한다.
#          그런데 SVM에서는 결정 경계를 정의하는 게 결국 서포트 벡터이기 때문에 데이터 포인트 중에서
#          서포트 벡터만 잘 골라내면 나머지 쓸 데 없는 수많은 데이터 포인트들을 무시할 수 있다.
#          그래서 매우 빠르다.
# kernel : 저차원의 공간을 고차원의 공간으로 매핑해주는 작업
# cost(C) : 얼마나 많은 데이터 샘플이 다른 클래스에 놓이는 것을 허용하는지를 결정.
# C값이 작을수록 많이 허용하고, 클수록 적게 허용
# C값을 낮게 설정하면 이상치들이 있을 가능성을 크게 잡아 일반적인 결정 경계를 찾아내고,
# 높게 설정하면 반대로 이상치의 존재 가능성을 작게 봐서 좀 더 세심하게 결정 경계를 찾아낸다.
# 따라서, C값은 결정 경계와 마진의 간격 결정.
# c가 높으면 마진이 좁아지고 train error가 작아진다. 그러나 새로운 데이터에서는 분류를 잘 함
# gamma(γ) : 한 train data당 영향을 미치는 범위 결정. 결정 경계를 얼마나 유연하게 그을 것인지.
# γ가 커지면 영향을 미치는 범위가 줄어들고, 결정 경계에 가까이 있는 데이터만이 선의 굴곡에 영향을 준다.
# 따라서 결정 경계가 구불구불하게 그어진다.(오버피팅 초래 가능)
# γ가 작아지면 데이터가 영향을 미치는 범위가 커지고, 대부분의 데이터가 결정 경계에 영향을 준다.
# 따라서 결정 경계가 직선에 가까워진다.(언더피팅 초래 가능)

svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      1.00      0.98        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       1.00      1.00      1.00        37
           5       0.93      1.00      0.97        28
           6       1.00      1.00      1.00        28
           7       1.00      1.00      1.00        33
           8       1.00      0.93      0.96        43
           9       1.00      0.97      0.98        32

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



### 4. SGD Classifier 

In [17]:
# Stochastic Gradient Descent Classifier(SGDClassifier)
# 배치 크기가 1인 경사하강법 알고리즘
# 즉, 확률적 경사하강법은 데이터 세트에서 무작위로 균일하게 선택한 하나의 예를 의존하여 각 단계의 예측 경사를 계산
# 배치? 경사하강법에서 단일 반복에서 기울기를 계산하는데 사용하는 예(data)의 총 개수

sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.83      0.93      0.88        42
           2       0.98      1.00      0.99        40
           3       0.87      0.97      0.92        34
           4       1.00      0.97      0.99        37
           5       0.96      0.96      0.96        28
           6       0.96      0.93      0.95        28
           7       0.94      0.97      0.96        33
           8       0.90      0.86      0.88        43
           9       0.96      0.75      0.84        32

    accuracy                           0.94       360
   macro avg       0.94      0.93      0.94       360
weighted avg       0.94      0.94      0.94       360



### 5. Logistic Regression

In [18]:
# Logistic Regression(선형회귀)
# 소프트맥스(softmas)함수를 사용한 다중 클래스 분류 알고리즘
# 이름은 회귀지만, 실제로는 분류를 수행

logistic_model = LogisticRegression(max_iter = 500, solver = 'liblinear')
# max_iter : epoch이랑 비슷. 얼마나 학습시킬것인지
# solver : 최적화 문제에 사용할 알고리즘. 데이터셋이 작은 경우 'liblinear'가 좋은 선택.
logistic_model.fit(X_train,y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.90      0.90      0.90        42
           2       0.97      0.97      0.97        40
           3       0.86      0.94      0.90        34
           4       0.95      0.97      0.96        37
           5       0.93      1.00      0.97        28
           6       0.96      0.93      0.95        28
           7       0.97      0.97      0.97        33
           8       0.93      0.88      0.90        43
           9       0.97      0.88      0.92        32

    accuracy                           0.94       360
   macro avg       0.95      0.95      0.94       360
weighted avg       0.95      0.94      0.94       360



### (6) 모델 평가
- 모델의 성능을 평가하는 지표로 무엇이 좋을까?
    - Accuracy
- sklearn.metrics에서 제공하는 평가지표 중 적절한 것을 선택하고, 선택한 이유 설명
    - digit 데이터셋의 경우 바르게 예측하는게 더 중요하기 때문에 평가지표로 'accuracy' 를 선택하였다. 
- 학습된 모델들의 테스트데이터 예측 결과를 어떻게 해석?
    - Decision Tree : 0.84
    - Random Forest : 0.96
    - SVM : 0.99
    - SGDClassifier: 0.93
    - Logistic Regression : 0.94
    ---
    - SVM 모델이 99%로 가장 높은 정확도를 보였다.
    - 따라서 손글씨의 숫자를 예측하는 문제는 SVM 모델이 가장 좋은 결과를 얻을 수 있다.

## 2. load_wine()

### (1) 필요한 모듈 import하기

In [19]:
from sklearn.datasets import load_wine

### (2) 데이터 준비

In [20]:
# 변수 wine에 데이터를 담아준다.
wine = load_wine()

### (3) 데이터 이해하기

In [21]:
# 데이터에 어떤 정보들이 담겨있는지 훓어보자!
wine

{'data': array([[1.423e+01, 1.710e+00, 2.430e+00, ..., 1.040e+00, 3.920e+00,
         1.065e+03],
        [1.320e+01, 1.780e+00, 2.140e+00, ..., 1.050e+00, 3.400e+00,
         1.050e+03],
        [1.316e+01, 2.360e+00, 2.670e+00, ..., 1.030e+00, 3.170e+00,
         1.185e+03],
        ...,
        [1.327e+01, 4.280e+00, 2.260e+00, ..., 5.900e-01, 1.560e+00,
         8.350e+02],
        [1.317e+01, 2.590e+00, 2.370e+00, ..., 6.000e-01, 1.620e+00,
         8.400e+02],
        [1.413e+01, 4.100e+00, 2.740e+00, ..., 6.100e-01, 1.600e+00,
         5.600e+02]]),
 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1

In [22]:
type(wine) # Bunch는 파이썬의 딕셔너리와 유사한 형테의 데이터 타입

sklearn.utils.Bunch

In [23]:
# wine data의 key값들을 불러와서 살펴보기
wine.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])

In [24]:
wine_data = wine.data

In [25]:
wine_data.shape 
# 총 178개의 데이터를 가지고 있고 각각 13개의 정보를 담고있다.

(178, 13)

In [26]:
# feature data 지정하기
wine_feature = wine.feature_names
wine_feature
# 13개의 정보는 'alcohol', 'malic_acid','ash'...와 같은 카테코리를 의미한다.

['alcohol',
 'malic_acid',
 'ash',
 'alcalinity_of_ash',
 'magnesium',
 'total_phenols',
 'flavanoids',
 'nonflavanoid_phenols',
 'proanthocyanins',
 'color_intensity',
 'hue',
 'od280/od315_of_diluted_wines',
 'proline']

In [27]:
# label data 지정하기
wine_label = wine.target
print(wine_label.shape)
wine_label
# label은 '0','1','2'값을 가진다.

(178,)


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2])

In [28]:
# target names 출력해보기
wine.target_names # 'class_0', 'class_1', 'class_2'으로 분류

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

In [29]:
wine_df = pd.DataFrame(data = wine_data, columns = wine.feature_names)
wine_df["label"] = wine.target
wine_df

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  label  
0                            3.92   1065.0      0  
1                            3.40   1050.0      0  
2                            3.17   1185.0      0  
3                            3.45   1480.0      0  
4                            2.93    735.0      0  
..                            ...      ...    ...  
173                          1.74    740.0      2  
174                          1.56    750.0      2  
175                          1.56    835.0      2  
176                          1.62    840.0      2  
177                          1.60    560.0      2  

[178 rows x 14 columns]

In [30]:
# 데이터 describe해보기
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

### (4) train, test 데이터 분리

In [31]:
X_train, X_test, y_train, y_test = train_test_split(wine_data,
                                                   wine_label,
                                                   test_size=0.2,
                                                   random_state=10)
print('X_train 개수: ',len(X_train),', X_test 개수: ', len(X_test))

X_train 개수:  142 , X_test 개수:  36


### (5) 다양한 모델로 학습시켜보기

### 1. Decision Tree

In [32]:
decision_tree = DecisionTreeClassifier(random_state=15)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.83      1.00      0.91        10
           1       1.00      0.78      0.88        18
           2       0.80      1.00      0.89         8

    accuracy                           0.89        36
   macro avg       0.88      0.93      0.89        36
weighted avg       0.91      0.89      0.89        36



### 2. Random Forest

In [33]:
random_forest = RandomForestClassifier(random_state=23)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.91      1.00      0.95        10
           1       1.00      0.83      0.91        18
           2       0.80      1.00      0.89         8

    accuracy                           0.92        36
   macro avg       0.90      0.94      0.92        36
weighted avg       0.93      0.92      0.92        36



### 3. SVM

In [34]:
svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.88      0.70      0.78        10
           1       0.81      0.72      0.76        18
           2       0.42      0.62      0.50         8

    accuracy                           0.69        36
   macro avg       0.70      0.68      0.68        36
weighted avg       0.74      0.69      0.71        36



### 4. SGD Classifier 

In [35]:
sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred, zero_division=1))

              precision    recall  f1-score   support

           0       1.00      0.60      0.75        10
           1       0.60      1.00      0.75        18
           2       1.00      0.00      0.00         8

    accuracy                           0.67        36
   macro avg       0.87      0.53      0.50        36
weighted avg       0.80      0.67      0.58        36



### 5. Logistic Regression

In [36]:
logistic_model = LogisticRegression(max_iter = 500, solver = 'liblinear')
# max_iter : epoch이랑 비슷. 얼마나 학습시킬것인지
# solver : 최적화 문제에 사용할 알고리즘. 데이터셋이 작은 경우 'liblinear'가 좋은 선택.
logistic_model.fit(X_train,y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        10
           1       1.00      0.83      0.91        18
           2       0.73      1.00      0.84         8

    accuracy                           0.92        36
   macro avg       0.91      0.94      0.92        36
weighted avg       0.94      0.92      0.92        36



### (6) 모델 평가
- 모델의 성능을 평가하는 지표로 무엇이 좋을까?
    - Accuracy
- sklearn.metrics에서 제공하는 평가지표 중 적절한 것을 선택하고, 선택한 이유 설명
    - 와인이 세 가지 종류 중에 무엇에 해당하는지 예측하는 것이다.
    - 이 경우, 전체 개수 중 맞은 것의 개수의 수치인 Accuracy값을 모델 성능 평가지표로 두었다.
- 학습된 모델들의 테스트데이터 예측 결과를 어떻게 해석?
    - Decision Tree : 0.89
    - Random Forest : 0.92
    - SVM : 0.69
    - SGDClassifier: 0.67
    - Logistic Regression : 0.92
    ---
    - Random Forest나 Logistic Regression이 92%의 가장 높은 정확도를 보였다.
    - 그러나 Random Forest는 선형 모델보다 많은 메모리를 사용하고 데이터가 많아지면 훈련과 예측이 느려지는 단점이 있다.
    - 따라서 와인의 종류를 분류하여 예측하는 문제는 Logistic Regression 모델로 예측하는 것이 좋다.

## 3. load_breast_cancer()

### (1) 필요한 모듈 import하기

In [37]:
from sklearn.datasets import load_breast_cancer

### (2) 데이터 준비

In [38]:
# 변수 cancer에 데이터를 담아준다.
cancer = load_breast_cancer()

### (3) 데이터 이해하기

In [39]:
# 데이터에 어떤 정보들이 담겨있는지 훓어보자!
cancer

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
         1.189e-01],
        [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
         8.902e-02],
        [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
         8.758e-02],
        ...,
        [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
         7.820e-02],
        [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
         1.240e-01],
        [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
         7.039e-02]]),
 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
        0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
        1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
        1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
        1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0

In [40]:
type(cancer) # Bunch는 파이썬의 딕셔너리와 유사한 형테의 데이터 타입

sklearn.utils.Bunch

In [41]:
# breast cancer data의 key값들을 불러와서 살펴보기
cancer.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

In [42]:
cancer_data = cancer.data

In [43]:
cancer_data.shape 
# 총 569개의 데이터를 가지고 있고 각각 30개의 정보를 담고있다.

(569, 30)

In [44]:
# feature data 지정하기
cancer_feature = cancer.feature_names
cancer_feature
# 30개의 정보는 'mean radius', 'mean texture','mean perimeter'...와 같은 카테코리를 의미한다.
# 여러 사람의 건강 지표에 대한 데이터 카테고리

array(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error',
       'fractal dimension error', 'worst radius', 'worst texture',
       'worst perimeter', 'worst area', 'worst smoothness',
       'worst compactness', 'worst concavity', 'worst concave points',
       'worst symmetry', 'worst fractal dimension'], dtype='<U23')

In [45]:
# label data 지정하기
cancer_label = cancer.target
print(cancer_label.shape)
cancer_label
# label은 '0','1'값을 가진다.
# 이때, '0'이 양성을 '1'이 음성을 뜻한다.

(569,)


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,

In [46]:
# target names 출력해보기
cancer.target_names # 'malignant', 'benign'으로 분류

array(['malignant', 'benign'], dtype='<U9')

In [47]:
cancer_df = pd.DataFrame(data = cancer_data, columns = cancer.feature_names)
cancer_df["label"] = cancer.target
cancer_df

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  label  
0                  0.2654          0.4601                  0.11890      0  
1                  0.1860          0.2750                  0.08902      0  
2                  0.2430          0.3613                  0.08758      0  
3                  0.2575          0.6638                  0.17300      0  
4                  0.1625          0.2364                  0.07678   

In [48]:
# 데이터 describe해보기
print(cancer.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

### (4) train, test 데이터 분리

In [49]:
X_train, X_test, y_train, y_test = train_test_split(cancer_data,
                                                   cancer_label,
                                                   test_size=0.2,
                                                   random_state=11)
print('X_train 개수: ',len(X_train),', X_test 개수: ', len(X_test))

X_train 개수:  455 , X_test 개수:  114


### (5) 다양한 모델로 학습시켜보기

### 1. Decision Tree

In [50]:
decision_tree = DecisionTreeClassifier(random_state=16)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.88      0.92      0.90        38
           1       0.96      0.93      0.95        76

    accuracy                           0.93       114
   macro avg       0.92      0.93      0.92       114
weighted avg       0.93      0.93      0.93       114



### 2. Random Forest

In [51]:
random_forest = RandomForestClassifier(random_state=24)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.95      0.97        38
           1       0.97      1.00      0.99        76

    accuracy                           0.98       114
   macro avg       0.99      0.97      0.98       114
weighted avg       0.98      0.98      0.98       114



### 3. SVM

In [52]:
svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.79      0.88        38
           1       0.90      1.00      0.95        76

    accuracy                           0.93       114
   macro avg       0.95      0.89      0.92       114
weighted avg       0.94      0.93      0.93       114



### 4. SGD Classifier 

In [53]:
sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred, zero_division=1))
# zero_division=1 넣고 안 넣고의 차이 발생 이유 찾기!!!

              precision    recall  f1-score   support

           0       0.97      0.87      0.92        38
           1       0.94      0.99      0.96        76

    accuracy                           0.95       114
   macro avg       0.95      0.93      0.94       114
weighted avg       0.95      0.95      0.95       114



### 5. Logistic Regression

In [54]:
logistic_model = LogisticRegression(max_iter = 500, solver = 'liblinear')
# max_iter : epoch이랑 비슷. 얼마나 학습시킬것인지
# solver : 최적화 문제에 사용할 알고리즘. 데이터셋이 작은 경우 'liblinear'가 좋은 선택.
logistic_model.fit(X_train,y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.94      0.89      0.92        38
           1       0.95      0.97      0.96        76

    accuracy                           0.95       114
   macro avg       0.95      0.93      0.94       114
weighted avg       0.95      0.95      0.95       114



### (6) 모델 평가
- 모델의 성능을 평가하는 지표로 무엇이 좋을까?
    - Recall
- sklearn.metrics에서 제공하는 평가지표 중 적절한 것을 선택하고, 선택한 이유 설명
    - 유방암 데이터셋의 경우, accuracy만으로 예측결과를 신뢰하기에는 문제가 있다.
    - 암환자를 찾아내는 것이 목표이기 때문에, recall값을 모델의 성능 지표로 두었다.
- 학습된 모델들의 테스트데이터 예측 결과를 어떻게 해석?
    - Decision Tree : 0.93
    - Random Forest : 0.98
    - SVM : 0.93
    - SGDClassifier: 0.94
    - Logistic Regression : 0.95
    ---
    - Random Forest 모델의 Recall값이 98%로 가장 높게 나왔다.
    - 따라서 Random Forest를 이용하여 유방암 여부를 예측하는 것이 가장 좋다.